File: cryptopals/adversary/average_hamming_distance.py
from collections.abc import Iterable
from itertools import combinations
from .. import Text, Block

# Number of blocks to compare to each other to find the most likely key length
# Higher number will be more accurate, but break on shorter ciphertexts
NUMBER_OF_BLOCKS: int = 6


# Not an adversary by itself, but used by both {brute_force_repeating_key_xor} and {find_aes_ecb_cipher}
def average_hamming_distance(text: Text) -> float:
    # This operation does not work if there are not enough blocks
    if (text.length / NUMBER_OF_BLOCKS) < text.block_size:
        raise ValueError('Invalid operation')

    # First {NUMBER_OF_BLOCKS} blocks of size {text.block_size}
    blocks: Iterable[Block] = (
        text.get_block(block_index)
        for block_index in range(NUMBER_OF_BLOCKS)
    )

    # Combine each block with each other block
    combos: list[tuple[Block, Block]] = list(combinations(blocks, 2))

    # Calculate the average hamming distance between blocks
    return sum(
        block_1.hamming_distance(block_2) / text.block_size
        for block_1, block_2 in combos
    ) / len(combos)

File: rust/src/adversary/average_hamming_distance.rs
use crate::{BlockIterator, CryptopalsError};

// Number of blocks to sample and compare to each other to find the most likely
// key length Higher number will be more accurate, but break on shorter
// ciphertexts
const NUMBER_OF_BLOCKS: usize = 6;

// C(n, r) = n! / [(n−r)! r!]
// C(6, 2) = 6! / [(6-2)! 2!] = 6 * 5 / 2 = 15
const NUMBER_OF_COMBINATIONS: f32 = 15.0;

/// Not an adversary by itself, but used by both
/// `attack_force_repeating_key_xor` and `detect_aes_ecb_cipher`
pub fn average_hamming_distance(iterator: BlockIterator) -> Result<f32, crate::CryptopalsError> {
    // Store block size before consuming iterator
    let block_size = iterator.block_size() as f32;

    // Get the first blocks and store in a Vec
    let blocks = iterator.take(NUMBER_OF_BLOCKS).collect::<Vec<_>>();

    // Create all possible combinations as a Tuple
    let combinations = (0..NUMBER_OF_BLOCKS)
        .flat_map(|i| (i..NUMBER_OF_BLOCKS).map(move |j| (i, j)))
        // Use indexes to get reference from Vec
        .map(|(index_1, index_2)| {
            let block_1 = blocks
                .get(index_1)
                .ok_or(CryptopalsError::NotEnoughBlocks)?;

            let block_2 = blocks
                .get(index_2)
                .ok_or(CryptopalsError::NotEnoughBlocks)?;

            Ok((block_1, block_2))
        })
        // Propagate Result::Err if there were not enough blocks
        .collect::<Result<Vec<_>, CryptopalsError>>()?;

    // Calculate total distance
    let total_distance = combinations
        .into_iter()
        // Calculate hamming distance for each combination
        .map(|(block_1, block_2)| block_1.hamming_distance(block_2))
        // Propagate Result::Err if any combination had unequal lengths
        .collect::<Result<Vec<_>, CryptopalsError>>()?
        // Turn back into iterator
        .into_iter()
        // Sum over all combination
        .sum::<usize>();

    Ok(total_distance as f32 / (block_size * NUMBER_OF_COMBINATIONS))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Bytes;

    #[test]
    fn just_enough_blocks() {
        let value = Bytes::from("Hello, World");
        let iter = value.block_iterator(2);

        let result = average_hamming_distance(iter).unwrap();

        assert_eq!(result, 43.0 / 15.0);
    }

    #[test]
    fn not_enough_blocks() {
        let value = Bytes::from("Hello, World");
        let iter = value.block_iterator(3);

        let error = average_hamming_distance(iter).unwrap_err();
        let expected = CryptopalsError::NotEnoughBlocks;

        assert_eq!(error, expected);
    }
}
